sklearn/preprocessing/_discretization.py

```diff
@@ -260,7 +276,14 @@ def transform(self, X):
         """
         check_is_fitted(self)
 
-        Xt = check_array(X, copy=True, dtype=FLOAT_DTYPES)
+        # check input and attribute dtypes
+        dtype = (np.float64, np.float32) if self.dtype is None else self.dtype
```

Comment:
Does this means that we should store `output_dtype` as a private attribute `self._output_dtype` Like this, we only do this inferring at `fit` time and don't do it again at `transform`.